### ISSUE TITLE
Enhancement: [prefer-literal-enum-member] Allow nested bitwise when using the `allowBitwiseExpressions` option

### ISSUE THREAD

**Issue by yeonjuan** [typescript-eslint/typescript-eslint#9166]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/play/#ts=5.2.2&showAST=es&fileType=.tsx&code=MYewdgzgLgBApmArgWxgNQIYBsCWATDKHcAMSwwHMIYBvAKBhgDlw4YBeGABgBo6GYAJTgBHRDgBOcQSBCxOARhgAeZdz6NhYyXAAiksBmRtFKtQo1DR4qQCEMEBEZMwlqmACZLWm3ACiAB5QhsYcrmYwAMze1joAKhIYOLhgFADKcAAOGIlQIBJhbmoALHwCAIJYWCAA7jJyhREArJaV1TX6EiEuRTAAbK1VtfaO3Y3uAOyD7YHBzuNqABzTtQlJKelZOYT5CzAAnCs1TBgAbjgUhMRgewq8FUM1AOrJeMA5eLcW-IwA9ABU-xgAAVCAALGDIRDQGAAIzYGBgp2w%2BBgeB0wDyEgAnjAJLJYP9fgJ6vIrNopKSYAAfGBtWpU2n0mprZI4VIZbK5fJlP6AkHgyHQ2DwmCIjCwiAgLCIKBsIkVSXS2UuHw6Rl0x4a5mdMZMx4jJyhfUzIJ6zXtVkbTnbLE0i21E7nS5EcC8mAAoGgqAQ5AYXGiqTkIinNi7CVSmVymAKzRwYM4UMAeQk5SVUZczO1j118xNwwcRrY%2BZqs3NzKt7M2XJ2BRLTouVzdPw9-O9vv9MHAWADCJgADNkkWY8TGIaxpw1XZC%2BXHmXnHwAL5AA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0Wlsfy8AhvHpNYAW1oTEEgEa90YANqRe0APbRIAGkyQR8DQHcAQpXzHKyRAFEAHhxTJKGpqgz44iAL4BdcDAfEB8gA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Description

As #9114 merged, it is allowed to use enum self-members in bitwise operations when the `allowBitwiseExpressions` option. However, if bit operations are used nested, a lint error occurs.
I think the rule should also allow when using nested bitwise operations if this option is set.

### Fail

```TypeScript
const OUT = 1 << 0,;
enum Foo {
   Bar = 1 << 1,
   Baz = OUT | Foo.Bar
}
```

### Pass

```TypeScript
const enum ValidationFlags {
  None = 0,

  RequireRoot = 1 << 0,
  RequireDirname = 1 << 1,
  RequireBasename = 1 << 2,
  RequireExtname = 1 << 3,
  RequireTrailingSeparator = 1 << 4,

  AllowRoot = 1 << 5,
  AllowDirname = 1 << 6,
  AllowBasename = 1 << 7,
  AllowExtname = 1 << 8,
  AllowTrailingSeparator = 1 << 9,
  AllowNavigation = 1 << 10,
  AllowWildcard = 1 << 11,

  /** Path must be a valid directory root */
  Root = RequireRoot | AllowRoot | AllowTrailingSeparator,

  /** Path must be a absolute */
  Absolute = RequireRoot | AllowRoot | AllowDirname | AllowBasename | AllowExtname | AllowTrailingSeparator | AllowNavigation,

  /** Path may be relative or absolute */
  RelativeOrAbsolute = AllowRoot | AllowDirname | AllowBasename | AllowExtname | AllowTrailingSeparator | AllowNavigation,

  /** Path may only be a filename */
  Basename = RequireBasename | AllowExtname,
}
```

### Additional Info

The pass example was found in the [typescript repository](https://github.com/microsoft/TypeScript/blob/79a851426c514a12a75b342e8dd2460ee6615f73/src/harness/vpathUtil.ts#L36-L63).

**Comment by kirkwaiblinger** [comment]
> My 2c - it's such a rare usecase to do something like this that I don't think it's worth the effort to introduce such an option.
> 
> Few codebases will use bitwise enums by themselves - let alone bitwise enums that are self-referencing.

@bradzacher There might be a misunderstanding? I think that's an argument against https://github.com/typescript-eslint/typescript-eslint/issues/7763, but not this issue. The option already exists (has just been created), but with a bug.

```ts
enum E {
  // The primary reason you would ever write these is to permit self-referential constructions like AandB or AandBandC
  A = 1 << 0,
  B = 1 << 1,
  C = 1 << 2,
  // This is allowed, and is the entire stated reason for https://github.com/typescript-eslint/typescript-eslint/issues/7763
  AandB = A | B,
  // This flags, but should be allowed, for the same reason as the previous one.
  AandBandC = A | B | C,
  ~~~~~~~~~
}
```

+1 to allowing arbitrary bit manipulations so long as the operands are numeric literals or other enum members. Numeric variables maybe should be allowed too to support masks?...but that seems problematic since it would require type information, which is hard to justify for a rule whose primary use case is purely syntactic.

**Comment by bradzacher** [comment]
I did misunderstand yeah.
But I still stand by what I said - it's still relevant!
We're in a very small intersection of codebases that need functionality like this.

**Comment by kirkwaiblinger** [comment]
Thinking more about this, I would amend my previous comment (again!), because I _can_ fathom a codebase that needs `1 << 0` (i.e. numeric literals) to work but does not need `A | B` (nonliterals) to work, since it would make sense to just set a bunch of independent enum members to different powers of two, without having any named combinations thereof, and then do bit manipulations at the usage sites for flags. In fact, (contrary to my previous assertion) this _may_ be the more prevalent usage, but I could see it going either way.

I _cannot_ fathom a codebase which needs `A | B` to work but wouldn't also expect `A | B | C` to work. 

Keep in mind, #9114, whose backing issue (#7763)'s express purpose (noted also @JoshuaKGoldberg's acceptance comment https://github.com/typescript-eslint/typescript-eslint/issues/7763#issuecomment-1771383669) is to support self-referential enum bit operations, was merged literally yesterday.  I think that issue is basically just not resolved at all without `A | B | C` support. If we don't include support for it, a different feature has been born, which is to support _non_ self referential flags, with incidental support for trivial self-referential flags (that I would consider a bug against that feature!).

Here's my personal thoughts, feel free to differ

| approach | satisfactory in Kirk's opinion? | effort to implement |
| --- | --- | --- |
| Leave as is | Not satisfactory; known half-baked feature on day 1 when it releases next Monday | Literally None |
| Remove the option altogether | Not satisfactory; we just merged it yesterday. (Would be a different story if it was a feature that had existed for a long time but no one cared about, and we decided to deprecate it) | Trivial |
| _Remove_ support for nonliterals in the bit operations | Acceptable; only ships a coherent, complete feature, albeit _not_ the one agreed to in issue request. (probably should also be reduced to only allowing `<<` for bitflags if we did go this route) | Slightly more effort, but still trivial |
| Add support for arbitrarily nested bit operations | Best outcome | Easy. Recurse on left and right operands of binary expressions. Marginally more difficult than it was for me to format this table |

**Comment by kirkwaiblinger** [comment]
Yeah, I don't have enough experience with using complicated flag enums to know whether there's a good use case for many other operators. It does seem like `1 << 2` and `A | B | C | D` are probably the vast majority of the use case. Feel like it wouldn't hurt to err on the side of being permissive rather than too restrictive either though. No strong feelings on the resolution to those details 🤷‍♂️ , except that deferring to whatever operands the rule currently supports seems least controversial.

**Comment by JoshuaKGoldberg** [comment]
I'd also +1 being generally permissive. I can see a use case where folks would want weird things like `(SetX & SetY) | SetZ` (or would it be `(SetX | SetY) & SetZ`?).

**Comment by JoshuaKGoldberg** [comment]
> wouldn't hurt to err on the side of being permissive rather than too restrictive

I'd also +1 being generally permissive. I can see a use case where folks would want weird things like `(SetX & SetY) | SetZ` (or would it be `(SetX | SetY) & SetZ`?).

@typescript-eslint/triage-team any objections to marking as accepting PRs?

**Comment by Josh-Cena** [comment]
There's a bug with this option. It should _not_ allow `A | B` as that defeats the purpose of this rule, which is to avoid the scoping gotchas of enums. I don't understand why the rule allows `C = A | B` but forbids `C = A`. If we only allow bitwise expressions of arbitrary depth, and all operands must be literals, then sure.

**Comment by kirkwaiblinger** [comment]
> There's a bug with this option. It should _not_ allow `A | B` as that defeats the purpose of this rule, which is to avoid the scoping gotchas of enums. I don't understand why the rule allows `C = A | B` but forbids `C = A`. If we only allow bitwise expressions of arbitrary depth, and all operands must be literals, then sure.

This is an intriguing comment, and I would be interested to engage on it in its own issue, but for now I don't see how it impacts/blocks _this_ issue. Could you clarify?

**Comment by Josh-Cena** [comment]
With `allowBitwiseExpressions`:

- `A` - banned
- `A | B` - should be banned but not, should be fixed
- `A | B | C` - should be banned and is, but this issue without being careful would allow it
- `0` - allowed
- `0 | 1` - allowed
- `0 | 1 | 2` - should be allowed and this issue proposes this

I'm saying we should fix the `A | B`, otherwise either this issue introduces _more_ bugs (as the motivation already said, it intends to use this option to support `AllowRoot | AllowDirname | AllowBasename | AllowExtname | AllowTrailingSeparator | AllowNavigation`, which I disagree should be supported (at least with this option)) and people will rely on it (it's a huge hole we are opening up than just allowing `A | B`), OR this issue introduces another inconsistency where `A | B` and `0 | 1 | 2` are both supported but `A | B | C` is not.

Who knows? Maybe if we fix this option to no longer allow non-literal bitwise operands, there turns out to be no good use case of nested bitwise, since the original intent of this option was only to support `1 << 4`!

**Comment by kirkwaiblinger** [comment]
I hear you, but I guess what I'm saying is

***

> `A | B` - should be banned but not, should be fixed

This is specifically, intentionally supported (a la https://github.com/typescript-eslint/typescript-eslint/issues/7763). I think that this requires a separate issue if we are to rethink that intention.

> `A | B | C` - should be banned and is, but this issue without being careful would allow it

That this should _not_ be banned is the whole point of this issue. It's a logical consequence of https://github.com/typescript-eslint/typescript-eslint/issues/7763 that this should also be supported if `A | B` is intentionally supported (which it _is_, unless https://github.com/typescript-eslint/typescript-eslint/issues/7763 is revisited).

> `A` - banned

IMO not 100% unambiguous in the way that `A | B | C` is; warrants separate issue.

***

I get that neither the `A` or `A | B | C` case would exist if https://github.com/typescript-eslint/typescript-eslint/issues/7763 is undone, but that's just not the case for now. I think we could label this issue as blocked by another issue if an issue to undo https://github.com/typescript-eslint/typescript-eslint/issues/7763 is created (and feel free to do so 🤷, though I'll say in advance I'll be -1 there), but, otherwise, this should continue.

**Comment by Josh-Cena** [comment]
Ah, I was not aware of https://github.com/typescript-eslint/typescript-eslint/issues/7763. Well, I think the `A` vs. `A | B` inconsistency is unfortunate, and I don't know what to do without splitting #7763 into its own option. Maybe we should make self-referential members behave fully identically to literals.

**Comment by JoshuaKGoldberg** [comment]
@Josh-Cena @kirkwaiblinger just confirming, it looks like there's consensus here? Are we good to accept PRs?

**Comment by Josh-Cena** [comment]
I'm okay to implement this issue in its current form just by virtue of #7763 being accepted and implemented :)
